### PR TITLE
 Add Grounded mobile sam for perception

### PR DESCRIPTION
**Pull request by yvsriram**
## Motivation and Context

- Created for sharing code

Other TODOs before merge:
- separate out evaluation code from perception code
- add GroundedSAM as submodule

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

# Checklist:

- [ ] I have made changes to existing documentation where needed.
- [ ] I have added tests that show that the PR is functional.
- [ ] I ran `tests/hw_manual_test.py` on hardware and observed that robot behavior looks normal.

### CODE REVIEW

File: src/home_robot/home_robot/perception/grounded_mobile_sam.py [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+                self.save_observations(obs, folder_name, method_name=method_name)
+            return self.quantitative_evaluate(folder_name, method_name=method_name)
+
+            # # evaluate object segmentation
+            # object_iou = self.evaluate(obs.semantic, gt_object_mask, gt_recep_mask)
+        except Exception as e:
+            print(f"Error: {e} for folder {folder_name}")
+            return None
+
+    def evaluate_folder(
```
Comment (cpaxton) [inline src/home_robot/home_robot/perception/grounded_mobile_sam.py:270]: should something like this just be in perception module?

Comment (cpaxton) [reply src/home_robot/home_robot/perception/grounded_mobile_sam.py:270]: this should not be sam specific

**Review by cpaxton — COMMENTED** [review]
minor comments

File: src/home_robot/home_robot/perception/grounded_mobile_sam.py [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+
+        # Building MobileSAM predictor
+        checkpoint = torch.load(MOBILE_SAM_CHECKPOINT_PATH)
+        self.mobile_sam = setup_model()
+        self.mobile_sam.load_state_dict(checkpoint, strict=True)
+        self.mobile_sam.to(device=DEVICE)
+        self.custom_vocabulary = custom_vocabulary
+        self.sam_predictor = SamPredictor(self.mobile_sam)
+
+    def reset_vocab(self, new_vocab: List[str]):
```
[the remark this answers is not in the document]
Comment (yvsriram) [reply src/home_robot/home_robot/perception/grounded_mobile_sam.py]: Copied from previous

File: src/home_robot/home_robot/perception/grounded_mobile_sam.py [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+        self.custom_vocabulary = new_vocab
+
+    def predict(self, obs: Observations, depth_threshold: Optional = None):
+        # Predict classes and hyper-param for GroundingDINO
```
[the remark this answers is not in the document]
Comment (yvsriram) [reply src/home_robot/home_robot/perception/grounded_mobile_sam.py]: Done

File: src/home_robot/home_robot/perception/grounded_mobile_sam.py [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+
+        obs.semantic = semantic_map.astype(int)
+        obs.task_observations = dict()
+        obs.task_observations["instance_map"] = instance_map
+        obs.task_observations["instance_classes"] = detections.class_id
+        obs.task_observations["instance_scores"] = detections.confidence
+        return obs
+
+    # Prompting SAM with detected boxes
+    def segment(self, image: np.ndarray, xyxy: np.ndarray) -> np.ndarray:
```
[the remark this answers is not in the document]
Comment (yvsriram) [reply src/home_robot/home_robot/perception/grounded_mobile_sam.py]: Done

File: src/home_robot/home_robot/perception/grounded_mobile_sam.py [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+        result_masks = []
+        for box in xyxy:
+            masks, scores, logits = self.sam_predictor.predict(
+                box=box, multimask_output=True
+            )
+            index = np.argmax(scores)
+            result_masks.append(masks[index])
+        return np.array(result_masks)
+
+    def read_images_from_folder(self, folder_name: str):
```
[the remark this answers is not in the document]
Comment (yvsriram) [reply src/home_robot/home_robot/perception/grounded_mobile_sam.py]: Removed this function

**Comment by yvsriram** [comment]
I removed some of the functions for standalone segmentation evaluation. I will add them back here, but this can be merged independent of that.

**Review by cpaxton — APPROVED** [review]
LGTM

**Pull request merged**